backend/dal/Services/Admin/Concrete/AgencyService.cs

```diff
@@ -100,35 +100,133 @@ public Agency Get(int id)
+                    updatedUsers.AddRange(users);
+                    this.Context.UpdateRange(users);
+                }
+                else if (agency.ParentId.HasValue)
+                {
+                    // Add the sub-agency to users who are associated with the parent.
+                    updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
+                }
+            }
 
-            this.Context.Entry(agency).CurrentValues.SetValues(entity);
-            base.Update(agency);
+            this.Context.Entry(original).CurrentValues.SetValues(agency);
+            base.Update(original);
+
+            // Mutate original entity.
+            this.Context.Entry(agency).CurrentValues.SetValues(original);
+            updatedUsers.Distinct().ForEach(u =>
+            {
+                this.Context.Entry(u).State = EntityState.Detached;
+                agency.Users.Add(new UserAgency(u, agency));
+            });
         }
 
         /// <summary>
         /// Remove the specified agency from the datasource.
         /// </summary>
         /// <exception type="KeyNotFoundException">Entity does not exist in the datasource.</exception>
-        /// <param name="entity"></param>
-        public override void Remove(Agency entity)
+        /// <param name="agency"></param>
+        public override void Remove(Agency agency)
         {
-            entity.ThrowIfNull(nameof(entity));
+            agency.ThrowIfNull(nameof(agency));
 
-            var agency = this.Context.Agencies.Find(entity.Id) ?? throw new KeyNotFoundException();
+            var original = this.Context.Agencies.Find(agency.Id) ?? throw new KeyNotFoundException();
+            var updateUsers = new List<User>();
 
-            this.Context.Entry(agency).CurrentValues.SetValues(entity);
-            base.Remove(agency);
+            // Any user associated with this agency needs to be updated.
+            if (original.ParentId.HasValue)
+            {
+                var users = this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId));
+                updateUsers.AddRange(users);
+            }
+            else
+            {
+                var users = this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(a => a.AgencyId == agency.Id));
+                users.ForEach(u => u.Agencies.Clear());
+                this.Context.Users.UpdateRange(users);
+                updateUsers.AddRange(users);
+            }
+
+            this.Context.Entry(original).CurrentValues.SetValues(agency);
+            base.Remove(original);
+
+            // Mutate original entity.
+            this.Context.Entry(agency).CurrentValues.SetValues(original);
+            updateUsers.ForEach(u =>
+            {
+                this.Context.Entry(u).State = EntityState.Detached;
+                agency.Users.Add(new UserAgency(u, agency));
```

Comment:
I don't think I understand what this is doing. If you detach the user and then add a UserAgency does that even modify the DB?

Reply:
Also, wouldn't this be adding a UserAgency for the agency that just got deleted?

Reply:
This code is ugly, but required either here or somewhere else.  If I leave the user attached to context then further operations on the user with existing code has conflicts regrettably.  In an ideal world any entity we returned from the DAL would be detached to ensure no unintentional tampering.
Adding the users to the agency is a simple way to return the users that were affected by the update.  

The best solution would probably be to give the DAL access to Keycloak...  But this also has complications and will take too much time.